### PR TITLE
Add support for ssm activations and managed instances

### PR DESCRIPTION
**Pull request by nickerickson**

### CODE REVIEW

File: c7n/resources/ssm.py [hunk 1]
```diff
@@ -16,21 +16,54 @@
...
-
     class resource_type(object):
-        service = 'ssm'
-        enum_spec = ('describe_parameters', 'Parameters', None)
-        name = "Name"
-        id = "Name"
-        filter_name = None
-        dimension = None
-        universal_taggable = True
+      service = 'ssm'
```
Comment (kapilt) [inline c7n/resources/ssm.py]: indentation here was changed to an inconsistent spacing.

File: c7n/resources/ssm.py [hunk 2]
```diff
@@ -16,21 +16,54 @@
...
+      id = "Name"
+      filter_name = None
+      dimension = None
+      universal_taggable = True
 
     retry = staticmethod(get_retry(('Throttled',)))
     permissions = ('ssm:GetParameters',
                    'ssm:DescribeParameters')
+
+@resources.register('ssm-managed-instances')
```
Comment (kapilt) [inline c7n/resources/ssm.py]: singular for resource names.

File: tests/test_ssm.py [hunk 3]
```diff
@@ -47,3 +47,60 @@ def test_ssm_parameter_not_secure(self):
         resources = p.run()
         self.assertEqual(len(resources), 1)
         self.addCleanup(client.delete_parameters, Names=['test-name', 'secure-test-name'])
+
+    @functional
```
Comment (kapilt) [inline tests/test_ssm.py]: this isn't a functional test, it depends on manually setup context that is not created as part of the test for the test to run. functional decorator should only be used for fully self-contained tests.

File: tests/test_ssm.py [hunk 4]
```diff
@@ -47,3 +47,60 @@ def test_ssm_parameter_not_secure(self):
...
+                  "value": True
+                }]
+            },
+            session_factory=session_factory,
+        )
+        resources = p.run()
+        self.assertEqual(len(resources), 2)
+
+    @functional
+    def test_ssm_activation_expired_no_registrations(self):
```
Comment (kapilt) [inline tests/test_ssm.py]: ditto re functional, but more to the point this test isn't adding any value additional to the previous test_ssm_activate_expired and should be removed.

File: tests/test_ssm.py [hunk 5]
```diff
@@ -47,3 +47,60 @@ def test_ssm_parameter_not_secure(self):
...
+                  "key": "RegistrationsCount",
+                  "value": 0
+                }]
+            },
+            session_factory=session_factory,
+        )
+        resources = p.run()
+        self.assertEqual(len(resources), 1)
+    
+    @functional
```
Comment (kapilt) [inline tests/test_ssm.py]: remove on the functional decorator.

File: tests/test_ssm.py [hunk 6]
```diff
@@ -47,3 +47,60 @@ def test_ssm_parameter_not_secure(self):
...
+        client = session_factory().client("ssm")
+        p = self.load_policy(
+            {
+                "name": "ssm-get-managed-instances",
+                "resource": "ssm-managed-instances"
+            },
+            session_factory=session_factory,
+        )
+        resources = p.run()
+        self.assertEqual(len(resources), 1)
```
Comment (kapilt) [inline tests/test_ssm.py]: please assert an actual value/attribute of the managed instance (instance-id) to verify that the test actual received a value that it expected.

File: c7n/resources/ssm.py [hunk 7]
```diff
@@ -16,21 +16,54 @@
...
+    permissions = ('ssm:DescribeInstanceInformation')
+
+@resources.register('ssm-activation')
+class SSMActivation(QueryResourceManager):
+    class resource_type(object):
+        service = 'ssm'
+        enum_spec = ('describe_activations', 'ActivationList', None)
+        id = 'ActivationId'
+        name = 'Description'
+        date = dimension = None
```
Comment (kapilt) [inline c7n/resources/ssm.py]: please update these values in the resource_type

```
date = 'CreatedDate'
filter_name = 'ActivationIds'
filter_type = 'list'
```

File: c7n/resources/ssm.py [hunk 8]
```diff
@@ -16,21 +16,54 @@
...
+        filter_name = None
+    permissions = ('ssm:DescribeActivations')
+
```
Comment (kapilt) [inline c7n/resources/ssm.py]: you need a trailing comma for this to be a valid single element string tuple.

File: c7n/resources/ssm.py [hunk 9]
```diff
@@ -16,21 +16,54 @@
...
+class ManagedInstance(QueryResourceManager):
+    class resource_type(object):
+        service = 'ssm'
+        enum_spec = ('describe_instance_information', 'InstanceInformationList', None)
+        id = 'InstanceId'
+        name = 'Name'
+        date = 'RegistrationDate'
+        dimension = None
+        filter_name = None
+    permissions = ('ssm:DescribeInstanceInformation')
```
Comment (kapilt) [inline c7n/resources/ssm.py]: you need a trailing comma for this to be a valid single element string tuple.

File: c7n/resources/ssm.py [hunk 10]
```diff
@@ -16,21 +16,54 @@
...
+@resources.register('ssm-managed-instances')
+class ManagedInstance(QueryResourceManager):
+    class resource_type(object):
+        service = 'ssm'
+        enum_spec = ('describe_instance_information', 'InstanceInformationList', None)
+        id = 'InstanceId'
+        name = 'Name'
+        date = 'RegistrationDate'
+        dimension = None
+        filter_name = None
```
Comment (kapilt) [inline c7n/resources/ssm.py]: this needs manual filter construction for use in serverless policies, ie. the api supports server side filtering but in a form thats not amenable to extant query support. We typically use a custom DescribeSource here, looks likes the same for all of the ssm resources, and we should just add query.py support for the form here, that can be separate from this pr though.

**Comment by nickerickson** [comment]
@kapilt updated to reflect all suggested changes minus the one for manual filter construction

File: c7n/resources/ssm.py [hunk 11]
```diff
@@ -34,3 +33,38 @@ class resource_type(object):
...
+    permissions = ('ssm:DescribeActivations',)
+
+@SSMActivation.action_registry.register('delete')
+class DeleteSSMActivation(Action):
+    schema = type_schema('delete')
+    permissions = ('ssm:DeleteActivation',)
+    def process(self, resources):
+        if not len(resources):
```
Comment (kapilt) [inline c7n/resources/ssm.py]: actions won't be invoked with an empty set.

**Comment by kapilt** [comment]
got a few lint issues, try `make lint` locally to see the report, on ci, you have to click through on travis to get to them.. copied inline below

```
c7n/resources/ssm.py:22:1: E302 expected 2 blank lines, found 1
c7n/resources/ssm.py:37:1: E302 expected 2 blank lines, found 1
c7n/resources/ssm.py:49:1: E302 expected 2 blank lines, found 1
c7n/resources/ssm.py:61:1: E302 expected 2 blank lines, found 1
c7n/resources/ssm.py:65:5: E301 expected 1 blank line, found 0
tests/test_ssm.py:53:9: F841 local variable 'client' is assigned to but never used
tests/test_ssm.py:59:19: E121 continuation line under-indented for hanging indent
tests/test_ssm.py:68:1: W293 blank line contains whitespace
tests/test_ssm.py:71:9: F841 local variable 'client' is assigned to but never used
tests/test_ssm.py:82:1: W391 blank line at end of file
make: *** [lint] Error 1```

**Comment by nickerickson** [comment]
@kapilt updated again with lint changes

**Pull request merged**